### ISSUE TITLE
QByteArrayData, QCharRef, QHashNode

### ISSUE THREAD

**Issue by TeeDeeY** [musescore/MuseScore#21819]
I am trying to build the Musescore executable from source

Source is 4.2.1, using Ubuntu 22.04 and Qt 6.6.2. 

Qt source files didn’t have QByteArrayData, QCharRef, QHashNode. The qt 5.1.5 source included them.  They contain a reference to another file.  I copied them from 5.1.5 into my 6.6.2 setup.  They passed that phase but had other data type errors. 

Is there a “correct” procedure to handle these three files? (Then maybe later compilation errors will be handled properly)

Thanks

**Comment by cbjeukendrup** [comment]
MuseScore 4.2.1 can only be built with Qt 5.15.x. 
Only the current master branch, which will become MuseScore 4.4, can be built using Qt 6.2.4.
Does that help?

**Comment by TeeDeeY** [comment]
Very helpful!  I’ll see what 4.2.4 looks like. 

Thank you.

**Comment by TeeDeeY** [comment]
For those using Qt6, in the CMakeLists.txt, set Qt5 compatibility to OFF on line 133.  It mostly compiled with 6.2.2 but didn’t finish.  I’ll get 6.2.4 shortly and see.  ;-)

**Comment by Jojo-Schmitz** [comment]
master __does__ fully build with Qt6, I'm doing it since a few weeks now

**Comment by TeeDeeY** [comment]
JoJo, thanks for the confirmation.  I reviewed some of the errors and they may be related to deprecation issues.  I’m using gcc on Ubuntu and may need to configure it with switches, or add #pragma statements.

**Comment by TeeDeeY** [comment]
The announcement said Qt 6.6.2 does not work.  I have to download 6.2.4, which was under the archive option.

**Comment by TeeDeeY** [comment]
I downloaded qt 6.2.4.  The earliest available qt6_design_studio is 4.1.0 lts.  When I use them together, I get unknown CMake command  _qt_internal_find_third_party_dependecies.  I used the archive option to get 6.2.4 but don’t have many options with the design studio.  

When I searched the internet, the error was supposedly due to mismatched Qt versions.  So, I was wondering which version on Ubuntu Linux is working for someone.

**Comment by Jojo-Schmitz** [comment]
You don't need Qt DesignStudio at all as far as I can tell.

**Comment by TeeDeeY** [comment]
I removed the environment variables pointing to qt studio files and changed findqt6.cmake.  I removed networkauth, core5compat and state machine.  Got through most of the compile but got an error looking for QOAuthOobReplyHandler.  I’ll try some other stuff or may just wait and try later   Maybe go with 5.15.

**Comment by Jojo-Schmitz** [comment]
You __do__ need networkauth, core5compat and state machine

**Comment by cbjeukendrup** [comment]
I recently updated our Wiki pages after the Qt 6 switch, see https://github.com/musescore/MuseScore/wiki/Install-Qt-and-Qt-Creator. If you follow these steps, things should work.

**Comment by TeeDeeY** [comment]
The updated pages work!
I re-downloaded the master zip file, downloaded only the essential qt files listed and only needed to use a build_overrides.cmake file with the QTDIR set. No other files needed to be modified. 

Thanks.

**Comment by TeeDeeY** [comment]
I compiled successfully using Ubuntu and tried running with -F.  The program crashes on note input. It is likely the libraries and plugins are not in the right locations.  Where are the locations?

The program at startup is trying to find files in $HOME/.local/share/MuseScore/MuseScore4Development/shortcuts.xml
… and other files 

Why are they looking in .local?  

I built using ./build.cmake clean install

From help files…
MuseScore fails to run

All necessary shared libraries are installed and in the right location.
Qt Plugins are installed and in the right location.

**Comment by cbjeukendrup** [comment]
It's expected that it looks for user configuration files in .local, and the absence of these files if you are launching musescore for the first time is also normal, and shouldn't cause a crash. 

Since the app does apparently launch correctly at first, I don't think it has to do with the locations of the libraries and plugins, because then it would probably not have launched at all. 

The most efficient way to find out why it crashes would be to use a debugger, and show the stack trace where it crashes.

**Comment by TeeDeeY** [comment]
OK.  I will try.  Strangely, I have built a qt6.2.4 version using Master and a qt5.15.2 with ms4.2.1.  They both crash when entering notes in a score. I’ll try the executables on a different computer and see what happens as well.

**Comment by cbjeukendrup** [comment]
It might have to do with playback, for example because of audio engine problems or a problem in Muse Sounds.

Also, are you using Qt Creator, or building from the command line? What command do you use the launch the executable? In the case of Qt Creator, are you sure you are running the executable from the install folder rather than the build folder?

**Comment by TeeDeeY** [comment]
I am not using qtcreator. I build using build.cmake from the same directory that the cmake file is in.  I tried both without parameters and also with “clean install” after reading the help areas. I run the executable by going into the subdirectory and running it as ./mscore      .  I’m in the directory itself.  In builds/Linux…/install/bin.  

I also tried ./build.cmake run.  It also runs the program, which crashed when I enter a note in the score.  I’ll try a different computer later and see if things change.  

I should try to make an app image but the instructions aren’t so clear.  I use app images on Ubuntu.

**Comment by TeeDeeY** [comment]
The mscore executable needs libraries when transferred to the other computer.  That will take time. 

It is likely as suggested a MuseSounds issue.  Using basic MS sounds, I can enter and play back drums and a cornettino.

**Comment by cbjeukendrup** [comment]
That's good to know. Then I would recommend to download the latest version of Muse Sounds Manager from https://musehub.com, and then make sure all Muse Sounds updates are installed. After doing that, you can verify the MuseSampler version within MuseScore by clicking Diagnostic > Muse Sampler > Check Muse Sampler. A dialog should appear, showing the Muse Sampler version, and that should be something like 0.5.1.

**Comment by TeeDeeY** [comment]
My version is already at 0.5.1 and sounds are all up-to-date because I recently updated them.  I'll uninstall, download and reinstall musehub just in case.  In the meantime, here are error messages.

The terminal describes the crash after I pressed the letter G. (also crashes when note specified entered via menu.  Rests don't cause a crash)

0:45:47.524 | INFO  | main_thread     | ActionsDispatcher::dispatch | try call action: note-g
Fatal (internal) error in ../celt/celt_decoder.c, line 118: assertion failed: st->mode == opus_custom_mode_create(48000, 960, NULL)
[3493:3493:20240308,204548.008447:ERROR elf_dynamic_array_reader.h:61] tag not found
[3493:3493:20240308,204548.010133:ERROR elf_dynamic_array_reader.h:61] tag not found
note: tag not found repeated multiple times before Aborted (core dumped) message.

**Comment by TeeDeeY** [comment]
Used dpkg-r muse-sounds-manager and then installed one that I just downloaded.  Issue is the same.  The program sees the version 0.5.1. It must be a system issue as both qt5 and qt6 crash on the same issue.

**Comment by TeeDeeY** [comment]
I got them both to run using “sudo ./mscore”.  So, the file locations need some investigation. 

Thanks for your help.

**Comment by TeeDeeY** [comment]
I will try changing audio codecs.  There is another post where that caused a crash in MuseScore.

**Comment by TeeDeeY** [comment]
Reinstalled Ubuntu 22.04.04 and ran the setup sh, installed libssl1.1 manually, and then did the system update.  It still crashes with MuseSounds. The error is related to opus.  

During build, messages says the opus CUSTOM_MODES is disabled.  It is in CMakeCache.txt but this gets overwritten during a clean build.  

How can I set it in build_overrides.cmake?

Thanks

**Comment by cbjeukendrup** [comment]
For what it's worth, it looks like there may be a connection between this issue and #21838.

**Comment by TeeDeeY** [comment]
My opus is 1.3 so it should be another problem.  I installed 22.04.04 and built a new setup.  It compiles but crashes on an opus error.  I’ll revisit in a week or so.  

Thanks

**Comment by TeeDeeY** [comment]
Does build.cmake use a precompiled opus library in the link phase?  I commented out the line 118 in celt_decode.c and did a ./build.cmake clean install run -F.  The same line 118 error showed up, making me think a library is being used rather that what I changed. 

Thanks

Getting CUSTOM_MODES to be defined was previously done but the celt_decode continues to process as if it were not defined and ran code inside an ifndef statement.

**Comment by cbjeukendrup** [comment]
I believe something complicated and annoying is going on: MuseScore itself uses opus from the source code that's in `thirdparty/opus` (MuseScore uses it for audio export), but MuseSampler uses it from (presumably) a system library.

**Comment by TeeDeeY** [comment]
Thanks for that insight.

**Comment by cbjeukendrup** [comment]
I don't really think it has to do with audio hardware, because both audio export and MuseSampler playback just render the audio using the CPU to some place in RAM, and as far as I know no special audio hardware is involved in that. (Only when MuseScore has mixed everything, the final result goes to the audio driver, but then Opus isn't involved anymore.)

**Comment by TeeDeeY** [comment]
Because the build works on other Ubuntu 22.04.04 systems, I was thinking the underlying hardware/driver could be an issue.  

I was speculating that the
opus_custom_mode_create may make an inquiry to the Ubuntu generic audio driver, to see if it is permissible.   My old hardware may not work well with the generic driver.  The “crash” is due to an assert check rather a coding error.  Just speculating.  Otherwise, failure is quite odd. 

The opus_custom_mode_create call does not ask for an error code and uses a NULL to not get one.  Getting an error code would make practical sense but it may be almost  impossible due to this being inside MuseSampler code.  

Is it possible to update an individual object file in the MuseSounds opus library? (Speculating dynamic linking could happen) Then, I would log the error code.

Thanks

**Comment by TeeDeeY** [comment]
I got it to run.  I had a mixture of items from various setup instructions. This was built from scratch. 

What works at the moment is usb stick install: Ubuntu 20.04 (older version)
During installation, do not download the latest updates as they may conflict with later setup packages. 
after installation, do NOT download Qt using the online installer qt-unified-Linux.  Later setup will get them. 
Download ci/Linux/setup.sh
Chmod +x setup.sh
./setup.sh
(This runs a long time. It creates a builds directory with stuff that is needed)
Download the libssl1.1 file and install with sudo dpkg per instructions.  I used version 2.22 since 2.17 wasn’t available. 
Chmod +x ~/build_tools/environment.sh
(For every terminal session you open, run the environment.sh until you put the paths in your build_overrides.cmake file)
.  ~/build_tools/environment.sh
Go to the Musescore directory and run ./build.cmake install run
(Should work.  Before I ran the environment.sh, I got an error on cmake.  I did an apt install cmake, but I probably didn’t have to).
